contrib/babelfishpg_tsql/src/session.c

```diff
@@ -376,3 +394,48 @@ initialize_context_table()
 
 	session_context_table = hash_create("Session Context", 128, &hash_options, HASH_ELEM | HASH_STRINGS);
 }
+
+/* 
+* This function is responsible for estimating the size of the entry and the number of keys 
+* and insert into the DSM for parallel workers
+* The first argument is ParallelContext which contains the info related to TOC
+* The second argument indicates whether we want to estimate the space or
+* we want to insert the data into DSM
+*/
+void
+babelfixedparallelstate_insert(ParallelContext *pcxt, bool estimate)
+{
+	BabelfishFixedParallelState *bfps;
+	int len;
+	char* current_db_name;
+	if (estimate)
+	{
+		/* Allow space to store the babelfish fixed-size parallel state. */
+		shm_toc_estimate_chunk(&pcxt->estimator, sizeof(BabelfishFixedParallelState));
+		shm_toc_estimate_keys(&pcxt->estimator, 1);
+	}
+	else
+	{
+		/* Initialize babelfish fixed-size state in shared memory. */
+		bfps = (BabelfishFixedParallelState *) shm_toc_allocate(pcxt->toc, sizeof(BabelfishFixedParallelState));
+		current_db_name = get_cur_db_name();
+		len = strlen(current_db_name);
+		strncpy(bfps->logical_db_name, current_db_name, MAX_BBF_NAMEDATALEND);
+		bfps->logical_db_name[len] = '\0';
+		shm_toc_insert(pcxt->toc, BABELFISH_PARALLEL_KEY_FIXED, bfps);
+		pfree(current_db_name);
+	}
+}
+
+/* This function is responsible for restoring the babelfixedparallelstate*/
+void
+babelfixedparallelstate_restore(shm_toc *toc)
+{
+	BabelfishFixedParallelState *bfps;	
+
+	/* Get the babelfish fixed parallel state from DSM */
+	bfps = shm_toc_lookup(toc, BABELFISH_PARALLEL_KEY_FIXED, false);
```

Comment:
If no key is found, then shm_toc_lookup returns NULL if 3rd argument is true, otherwise throws elog(ERROR). As we are setting the 3rd arg as FALSE, if the key is not found, we will get an error like
```
could not find key " <key> " in shm TOC at <toc>
```